lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,308 @@
+
+        # Extract data array dimensions to define output arrays.
+        dataarray_shape = next(temperature_cube.slices(
+            ["realization", y_coord, x_coord])).data[0].shape
+        dataarray_size = dataarray_shape[0] * dataarray_shape[1]
+
+        # Array containing all of the subsections extracted from data array.
+        all_temp_subsections = np.zeros((dataarray_size, footprintarray_size))
+        all_orog_subsections = np.zeros((dataarray_size, footprintarray_size))
+
+        # Creates cube list to hold lapse rate data.
+        lapse_rate_cube_list = iris.cube.CubeList([])
+
+        slices_over_realization = temperature_cube.slices_over("realization")
+        for temp_slice in slices_over_realization:
+
+            # Create slice to store lapse rate values.
+            lapse_rate_slice = temp_slice
+
+            temperature_data = temp_slice.data
+
+            # Fill sea points with NaN values. Can't use Numpy mask since not
+            # recognised by "generic_filter" function.
+            temperature_data = np.where(land_sea_mask, temperature_data,
+                                        np.nan)
+
+            # cval is value given to points outside the array.
+            fnc = fnc_class(allbuffers=all_temp_subsections)
+            generic_filter(temperature_data, fnc.filter, size=footprint_len,
+                           mode='constant', cval=np.nan)
+
+            fnc = fnc_class(allbuffers=all_orog_subsections)
+            generic_filter(orography_data, fnc.filter, size=footprint_len,
+                           mode='constant', cval=np.nan)
+
+            # height_diff_mask is True for points where the height
+            # difference between the central point and its neighbours
+            # is > max_height_diff.
+            height_diff_mask = create_height_diff_mask(
+                all_orog_subsections, ind_central_point, max_height_diff)
+
+            # Mask points with extreme height differences as NaN.
+            all_orog_subsections = np.where(height_diff_mask, np.nan,
+                                            all_orog_subsections)
+            all_temp_subsections = np.where(height_diff_mask, np.nan,
+                                            all_temp_subsections)
+
+            # Loop through both arrays and find gradient of each subsection.
+            # The gradient indicates lapse rate - save into another array.
+            lapse_rate_array = np.zeros(dataarray_size)
+
+            # TODO: This for loop is the bottleneck in the code - 1 array takes
+            # approximately 29 seconds out of a total running time of 33
+            # seconds for UKVX data. This section needs to be parallelised.
+            for i in range(dataarray_size):
+                lapse_rate_array[i] = calc_lapse_rate(
+                    all_temp_subsections[i, :], all_orog_subsections[i, :],
+                    ind_central_point)
```

Comment:
This plugin makes clever use of the scipy generic filter to try and vectorize the whole process, but it's not at all obvious how we bypass this step. Almost inevitably this will be something to throw out to a large multiprocessing pool to get through it quickly. If this is the case we may have to ask the question of whether we could have a much simpler code that requires less memory and just performs all the steps inside a multiprocess loop; if we've got to go point-by-point, why not do it all whilst we are in that loop?

Reply:
Fair point:
1) I think one downside to a loop that goes through each point is that it would require a nested loop to iterate over the x/y dimensions. While this code still has one loop - its just the one. So there's a code speed up there. In addition - the rows of data in this one loop don't have to be calculated sequentially which would make future parallelisation simpler.
2) The generic filter function takes care of what to do with the borders of the dataset, whereas a nested loop needs additional statements to change its behaviour at the borders.